Add docs and template ProviderDataIngester

stacimc opened the pull request:
## Fixes
<!-- If PR doesn't fully resolve the issue, replace 'Fixes' below with 'Related to'. -->
<!-- If there is no issue being resolved, please consider opening one before creating this pull request. -->
Fixes #757 by @stacimc

## Description
<!-- Concisely describe what the pull request does. -->
<!-- Add screenshots, videos, or other media to show the problem and the solution when appropriate. -->
* Adds a `create_provider_ingester` script that can be used to generate a templated provider api script and test file. The generated files include a lot of documentation and TODOs aimed at helping a new contributor flesh out an ingestion class. Credit to an [earlier template script in this repository](https://github.com/WordPress/openverse-catalog/tree/1c9fd4cf46f7c90f8e9db6eaacb31d6ad031fcf5/openverse_catalog/templates), from which I lifted quite a bit of code!
* Adds the following docs (would love better title suggestions 😅):
  * `adding_a_new_provider.md`: Details a very brief overview of what it even means to add a provider to Openverse, explains how to use the script, and also a brief explanation of how to add a `ProviderWorkflow` to generate the actual DAG
  * `provider_data_ingester_faq.md`: Describes "advanced options" for implementing some common non-standard use cases in a ProviderDataIngester, framed as an FAQ
  * `data_models.md`: Extremely temporary documentation of our columns, meant to be replaced by #783.
* Expands some of the documentation in the actual `ProviderDataIngester` class

The intention is to avoid too much duplication of documentation from the _code_. If deemed necessary, we can look into auto-generating some from the doc strings, but my strategy was:
* Script makes it easy to generate a stubbed out new ingester class
* Comments and TODOs within the generated class lead the developer through implementing a "straightforward" provider
* "Advanced" options are documented separately with examples

There's definitely a balancing act of where to document things in the code vs the docfile vs the template. My goal is mostly for this to be a good starting point!

## Testing Instructions
<!-- Give steps for the reviewer to verify that this PR fixes the problem; or delete this section entirely. -->

* Read the docs , make sure links work, etc :)
* Test the new script
  * `just test`
    * Verify that the test files (`foobar_industries.py` & `test_foobar_industries.py`) were deleted after the tests finished
  * Run it with some sample data, ie: `just add-provider "Foo Museum" "https://foo-museum.org/api/v1/" image audio`
    * Verify the `foo_museum.py` provider script and `test_foo_museum.py` files were generated
    * Inspect the files to make sure they make sense and that the name/endpoint/media types were templated in properly
    * Try again with different media types, including a single media type
    

## Checklist
<!-- Replace  the [ ] with [x] to check the boxes. -->
- [ ] My pull request has a descriptive title (not a vague title like `Update index.md`).
- [ ] My pull request targets the *default* branch of the repository (`main`) or a parent feature branch.
- [ ] My commit messages follow [best practices][best_practices].
- [ ] My code follows the established code style of the repository.
- [ ] I added or updated tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [ ] I tried running the project locally and verified that there are no visible errors.

[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin
<!-- You must read and understand the following attestation. -->

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

krysal commented on openverse_catalog/dags/templates/create_provider_ingester.py:
```diff
@@ -0,0 +1,152 @@
+    args = parser.parse_args()
+    provider = args.provider
+    endpoint = args.endpoint
+
+    # Get valid media types
+    media_types = []
+    for media_type in args.media:
+        if media_type in MEDIA_TYPES:
+            media_types.append(media_type)
+        else:
+            print(f"Ignoring invalid type {media_type}")
+
+    # Default to image if no valid media types given
+    if not media_types:
+        print('No media type given, defaulting to ["image",]')
+        media_types = [
+            "image",
+        ]
```
When I tried the script without specifying the media type I got an error. This way should be fixed to `image` as default:

```suggestion
    media_types = []

    # Default to image if no valid media types given
    if not args.media:
        print('No media type given, defaulting to ["image",]')
        media_types = [
            "image",
        ]
    else:
        # Get valid media types
        for media_type in args.media:
            if media_type in MEDIA_TYPES:
                media_types.append(media_type)
            else:
                print(f"Ignoring invalid type {media_type}")
```

stacimc replied:
Thank you for catching this! I've pulled this out into a separate method and unit tested it as well :)

krysal reviewed (commented):
This is fantastic 🤩 I started using the template to convert the Smithsonian script so I found some minor typos and details but this is looking **so good and useful**.

Thank you for writing such good documentation @stacimc! I'll get back after looking at the test files.

zackkrida commented on openverse_catalog/docs/adding_a_new_provider.md:
```diff
@@ -0,0 +1,90 @@
+
+## Overview
+
+The Openverse Catalog collects data from the APIs of sites that share openly-licensed media,and saves them in our Catalog database. This process is automated by Airflow DAGs generated for each provider. A simple provider DAG looks like this:
```
```suggestion
The Openverse Catalog collects data from the APIs of sites that share openly-licensed media, and saves them in our Catalog database. This process is automated by Airflow DAGs generated for each provider. A simple provider DAG looks like this:
```

AetherUnbound replied:
Small spacing fix, and also it might be good to reference Airflow documentation here:
```suggestion
The Openverse Catalog collects data from the APIs of sites that share openly-licensed media, and saves them in our Catalog database. This process is automated by [Airflow DAGs](https://airflow.apache.org/docs/apache-airflow/stable/concepts/dags.html) generated for each provider. A simple provider DAG looks like this:
```

zackkrida commented on openverse_catalog/docs/adding_a_new_provider.md:
```diff
@@ -0,0 +1,90 @@
+
+We call the code that pulls data from our provider APIs "Provider API scripts". You can find examples in [`provider_api_scripts` folder](../dags/providers/provider_api_scripts). This code will be run during the `pull` steps of the provider DAG.
+
+At a high level, a provider script should iteratively request batches of records from the provider API, extract data in the format required by Openverse, and commit it to local storage. Much of this logic is implemented in a [`ProviderDataIngester` base class](../dags/providers/provider_api_scripts/provider_data_ingester.py) (which also provides additional testing features *<TODO: link to documen ...
```
```suggestion
At a high level, a provider script should iteratively request batches of records from the provider API, extract data in the format required by Openverse, and commit it to local storage. Much of this logic is implemented in a [`ProviderDataIngester` base class](../dags/providers/provider_api_scripts/provider_data_ingester.py) (which also provides additional testing features *<TODO: link to documentation for testing features like ingestion_limit, skip_ingestion_errors etc>*). To add a new provider, extend this class and implement its abstract methods.
```

To remove any claims about simplicity. This is only one possible alternative.

stacimc replied:
Good point 😅

zackkrida reviewed (commented):
We might want to consider sanitizing the provided provider name in these scripts, or alternatively excluding unsupported characters. In my test I used the following configuration:

```
python3 openverse_catalog/dags/templates/create_provider_ingester.py "Nappy.co" "https://api.nappy.co/v1/openverse/images" -m "image"`
```

Which resulted in `nappy.co.py` filenames and a `class Nappy.coDataIngester`.

zackkrida commented on openverse_catalog/dags/templates/template_provider.py_template:
```diff
@@ -0,0 +1,148 @@
+        }
+
+
+def main():
+    # Allows running ingestion from the CLI without Airflow running for debugging
+    # purposes.
+    logger.info("Begin: {provider} data ingestion")
+    ingester = {provider}DataIngester()
+    ingester.ingest_records()
```
Is there a `just` recipe we could add to make this easier? As a non-python-person I don't follow how to actually try and run this without airflow.

zackkrida commented on openverse_catalog/dags/templates/template_provider.py_template:
```diff
@@ -0,0 +1,148 @@
+import logging
+
+from common import constants
+from common.license import get_license_info
```
```suggestion
from common.licenses import get_license_info
```

Small typo here.

zackkrida reviewed (commented):
@stacimc, this is amazing! I was able to use it to write my first DAG (and first python commit of my life, I believe 😅). Other than my inline comments, one thing I noticed was that by using 

```
python3 openverse_catalog/dags/templates/create_provider_ingester.py "Nappy" "https://api.nappy.co/v1/openverse/images" -m "image"
```

it created a constant called `NAPPY_IMAGE_PROVIDER` and I was only able to get things working by renaming to `NAPPY_DEFAULT_PROVIDER`.

Edit: The above might actually be wrong. It may be because the docs comment is required at the top of the provider script and isn't part of the template. Here's the example of what I added to the Nappy pr:

```
"""
Content Provider:       Nappy

ETL Process:            Use the API to identify all CC0-licensed images.

Output:                 TSV file containing the image meta-data.

Notes:                  This api was written specially for Openverse.
                        There are no known limits or restrictions.

"""
```

I don't feel qualified to fully review the PR but this is really awesome!

zackkrida commented:
One other thought. As I recall `get_should_continue` wasn't included in the template. That seems common enough to me that perhaps it could be included, but commented out?

stacimc commented:
> We might want to consider sanitizing the provided provider name in these scripts,

Added some extra sanitization to the provider name. It now replaces both spaces and periods with underscores, and then removes excluded chars (anything non-alphanumeric except hyphen and underscore). I think underscore replacement makes sense for period.

> it created a constant called NAPPY_IMAGE_PROVIDER and I was only able to get things working by renaming to NAPPY_DEFAULT_PROVIDER.

> Edit: The above might actually be wrong. It may be because the docs comment is required at the top of the provider script and isn't part of the template. Here's the example of what I added to the Nappy pr:

@zackkrida I'm not sure what's happening here! I don't think the constant name should be an issue, since as noted in the TODOs you need to actually define the constant yourself in `provider_details.py`. The docs comment you reference also shouldn't be required (our other refactored DAGs don't have that comment)[^1].

I tried playing around with this locally with a fake provider and it all works fine for me. Did you run `just up`?

[^1]: The fact that our refactored DAGs don't have this string is actually a problem though! We use it to generate our DAG documentation. I've created #803 to make sure we add it back in, and I'll go add it to the template now too 😄 I'm really glad you mentioned it!

stacimc commented on openverse_catalog/docs/provider_data_ingester_faq.md:
````diff
@@ -0,0 +1,142 @@
+
+**Solution**: In this case, you can reuse the `get_response_json` method by passing in the endpoint you need:
+
+```
````
Excellent point, that's a huge improvement! Collapsing the rest of these into a single commit :)

krysal reviewed (approved):
These docs and templates are very helpful. Great work here! ⭐️

I found some minor non-blocking observations, tiny details.

AetherUnbound commented on openverse_catalog/docs/adding_a_new_provider.md:
```diff
@@ -0,0 +1,90 @@
+At a high level the steps are:
+
+1. `generate_filename`: Generates a TSV filename used in later steps
+2. `pull_data`: Actually pulls records from the provider API, collects just the data we need, and commits it to local storage in TSVs.
```
```suggestion
2. `pull_data`: Pulls records from the provider API, collects just the data we need, and commits it to local storage in TSVs.
```

AetherUnbound commented on openverse_catalog/docs/adding_a_new_provider.md:
```diff
@@ -0,0 +1,90 @@
+
+1. `generate_filename`: Generates a TSV filename used in later steps
+2. `pull_data`: Actually pulls records from the provider API, collects just the data we need, and commits it to local storage in TSVs.
+3. `load_data`: Loads the data from TSVs into the actual Catalog database, updating old records and discarding duplicates.
```
```suggestion
3. `load_data`: Loads the data from TSVs into the Catalog database, updating old records and discarding duplicates.
```

stacimc replied:
Bad habit!

AetherUnbound commented on openverse_catalog/docs/adding_a_new_provider.md:
```diff
@@ -0,0 +1,90 @@
+Now that you have an ingester class, you're ready to wire up a provider DAG in Airflow to automatically pull data and load it into our Catalog database. This is as simple as defining a `ProviderWorkflow` configuration dataclass and adding it to the `PROVIDER_WORKFLOWS` list in [`provider_workflows.py`](../dags/providers/provider_workflows.py). Our DAG factories will pick up the configuration and  ...
+
+At minimum, you'll need to provide the following in your configuration:
+* `provider_script`: the name of the file where you defined your `ProviderDataIngester` class
+* `ingestion_callable`: the `ProviderDataIngester` class itself
+* `media_types`: the media types your provider handles
```
I'm realizing that we should make an issue for removing the legacy wrapper handling logic, which also means that we should be able to remove `provider_script` here eventually. I can make that issue 🙂 Also, I think we could even determine `media_types` automatically by using `ProviderDataIngester::providers.keys()` 😮 Lots of pairing down we could do on these configs once the refactors are all done! 🥳

AetherUnbound commented on openverse_catalog/docs/adding_a_new_provider.md:
```diff
@@ -0,0 +1,90 @@
+
+There are many other options that allow you to tweak the `schedule` (when and how often your DAG is run), timeouts for individual steps of the DAG, and more. These are documented in the definition of the `ProviderWorkflow` dataclass. *<TODO: add docs for other options.>*
+
+After adding your configuration, run `just up` and you should now have a fully functioning provider DAG! *<TODO: add and link to docs for how to run provider DAGs locally, preferably with images.>* *NOTE*: when your code is merged, the DAG will become available in production but will be disabled by default. A contributor with Airflow access will need to manually turn the DAG on in production.
```
I would split this across lines, it seems like markdown doesn't render the `* *` very well:
```suggestion
After adding your configuration, run `just up` and you should now have a fully functioning provider DAG! *<TODO: add and link to docs for how to run provider DAGs locally, preferably with images.>*

**NOTE**: when your code is merged, the DAG will become available in production but will be disabled by default. A contributor with Airflow access will need to manually turn the DAG on in production.
```

AetherUnbound commented on openverse_catalog/docs/data_models.md:
```diff
@@ -0,0 +1,57 @@
+TODO: This documentation is temporary and should be replaced by more thorough documentation of our DB fields in https://github.com/WordPress/openverse-catalog/issues/783>
```
Potentially?
```suggestion
*<TODO: This documentation is temporary and should be replaced by more thorough documentation of our DB fields in https://github.com/WordPress/openverse-catalog/issues/783>*
```

AetherUnbound commented on openverse_catalog/docs/data_models.md:
```diff
@@ -0,0 +1,57 @@
+| field name | description |
+| --- | --- |
+| *foreign_landing_url* | URL of page where the record lives on the source website. |
+| *audio_url* / *image_url* | Direct link to the media file. Note that the field name differs depending on media type. |
```
```suggestion
| *audio_url* / *image_url* | Direct link to the media file. Note that until [issue #784 is addressed](https://github.com/WordPress/openverse-catalog/issues/784) the field name differs depending on media type. |
```

AetherUnbound commented on openverse_catalog/docs/data_models.md:
```diff
@@ -0,0 +1,57 @@
+| --- | --- |
+| *foreign_landing_url* | URL of page where the record lives on the source website. |
+| *audio_url* / *image_url* | Direct link to the media file. Note that the field name differs depending on media type. |
+| *license_info* | LicenseInfo object that has (1) the URL of the license for the record, (2) string representation of the license, (3) version of the license, (4) raw license URL that was by provider, if different from canonical URL |
```
Some suggestions here with code references:
```suggestion
| *license_info* | [LicenseInfo object](https://github.com/WordPress/openverse-catalog/blob/8423590fd86a0a3272ca91bc11f2f37979048181/openverse_catalog/dags/common/licenses/licenses.py#L25) that has (1) the URL of the license for the record, (2) string representation of the license, (3) version of the license, (4) raw license URL that was by provider, if different from canonical URL. Usually generated by calling [`get_license_info`](https://github.com/WordPress/openverse-catalog/blob/8423590fd86a0a3272ca91bc11f2f37979048181/openverse_catalog/dags/common/licenses/licenses.py#L29) on respective fields returns/available from the API. |
```

AetherUnbound commented on openverse_catalog/docs/data_models.md:
```diff
@@ -0,0 +1,57 @@
+| *duration* | Audio duration in milliseconds. |
+| *bit_rate* | Audio bit rate as int. |
+| *sample_rate* | Audio sample rate as int. |
+| *category* | Category such as 'music', 'sound', 'audio_book', or 'podcast'. |
```
I believe this is available to all media types. Can we add a reference to it in the image-specific fields as well? Also it might be worth referencing the category enums we have: https://github.com/WordPress/openverse-catalog/blob/8423590fd86a0a3272ca91bc11f2f37979048181/openverse_catalog/dags/common/loader/provider_details.py#L118

(or, it looks like just enum? 🤔 doesn't seem like we have an audio category enum, I'll make an issue for that)

stacimc replied:
Definitely! This documentation at the moment is basically just copying docstrings from the image stores. This doc in particular should be totally rewritten in the Data normalization project.

obulat replied:
I've left a lot of comments before I got to this one, @stacimc  :) Totally fine if we leave everything as is before the re-write. Except for the `foreign_identifier` being required.

AetherUnbound commented on openverse_catalog/docs/data_models.md:
```diff
@@ -0,0 +1,57 @@
+| *set_position* | Position of the audio in the audio_set. |
+| *set_thumbnail* | URL of the audio_set thumbnail. |
+| *set_url* | URL of the audio_set. |
+| *alt_files* | A dictionary with information about alternative files for the audio (different formats/quality). Dict should have the following keys: url, filesize, bit_rate, sample_rate.
```
```suggestion
| *alt_files* | A dictionary with information about alternative files for the audio (different formats/quality). Dict should have the following keys: *url*, *filesize*, *bit_rate*, *sample_rate*.
```

AetherUnbound commented on openverse_catalog/docs/provider_data_ingester_faq.md:
````diff
@@ -0,0 +1,142 @@
+    ...
+```
+
+When doing this, keep in mind that adding too many requests may slow down ingestion. Be aware of rate limits from your provider API as well.
````
This is a really important point, so I think extra emphasis is worth it:
```suggestion
**NOTE**: When doing this, keep in mind that adding too many requests may slow down ingestion. Be aware of rate limits from your provider API as well.
```

AetherUnbound commented on openverse_catalog/docs/provider_data_ingester_faq.md:
```diff
@@ -0,0 +1,142 @@
+
+## What if my API endpoint isn't static and needs to change from one request to another?
+
+Example: Rather than passing a `page` number in query parameters, a provider expects the `page` as part of the endpoint path itself.
```
```suggestion
**Example**: Rather than passing a `page` number in query parameters, a provider expects the `page` as part of the endpoint path itself.
```

AetherUnbound commented on openverse_catalog/docs/provider_data_ingester_faq.md:
```diff
@@ -0,0 +1,142 @@
+
+Example: Rather than passing a `page` number in query parameters, a provider expects the `page` as part of the endpoint path itself.
+
+If your `endpoint` needs to change, you can implement it as a `property`:
```
```suggestion
**Solution**: If your `endpoint` needs to change, you can implement it as a `property`:
```

AetherUnbound commented on openverse_catalog/docs/provider_data_ingester_faq.md:
````diff
@@ -0,0 +1,142 @@
+
+```python
+def __init__(self, *args, **kwargs):
+    # IMPORTANT!
````
🚲 🏠 for sure, but maybe this would convey the point a bit more explicitly?
```suggestion
    # REQUIRED!
```

AetherUnbound commented on openverse_catalog/docs/provider_data_ingester_faq.md:
```diff
@@ -0,0 +1,142 @@
+
+## How do I run ingestion for a set of discrete categories?
+
+**Example**: My provider has some set of categories that I'd like to iterate over and ingest data for. Eg, an audio provider's search endpoint that requires you specify whether you're searching for "podcasts", "music", etc. I'd like to iterate over all the available categories and run ingestion for each.
```
Nit:
```suggestion
**Example**: My provider has some set of categories that I'd like to iterate over and ingest data for. E.g., an audio provider's search endpoint requires you specify whether you're searching for "podcasts", "music", etc. I'd like to iterate over all the available categories and run ingestion for each.
```

AetherUnbound replied:
Ah! That's fair 🙂

AetherUnbound commented on openverse_catalog/templates/create_provider_ingester.py:
```diff
@@ -0,0 +1,180 @@
+def _get_filled_template(
+    template_path: Path, provider: str, endpoint: str, media_types: list[str]
+):
+    with template_path.open("r", encoding="utf-8") as template:
+        camel_provider = inflection.camelize(provider)
+        screaming_snake_provider = inflection.underscore(provider).upper()
+
+        # Build provider configuration
+        provider_configuration = "\n        ".join(
+            _render_provider_configuration(screaming_snake_provider, media_type)
+            for media_type in media_types
+        )
+
+        template_string = template.read()
+        script_string = (
+            template_string.replace("{provider}", camel_provider)
+            .replace("{screaming_snake_provider}", screaming_snake_provider)
+            .replace("{provider_underscore}", inflection.underscore(provider))
+            .replace("{provider_data_ingester}", f"{camel_provider}DataIngester")
+            .replace("{endpoint}", endpoint)
+            .replace("{provider_configuration}", provider_configuration)
+        )
+
+        return script_string
```
UTF-8 is the default encoding, "read" mode is the default mode, and [`Path::read_text`](https://docs.python.org/3/library/pathlib.html#pathlib.Path.read_text) simplifies this a lot, so this could be:

```suggestion
    camel_provider = inflection.camelize(provider)
    screaming_snake_provider = inflection.underscore(provider).upper()

    # Build provider configuration
    provider_configuration = "\n        ".join(
        _render_provider_configuration(screaming_snake_provider, media_type)
        for media_type in media_types
    )

    template_string = template_path.read_text()
    script_string = (
        template_string.replace("{provider}", camel_provider)
        .replace("{screaming_snake_provider}", screaming_snake_provider)
        .replace("{provider_underscore}", inflection.underscore(provider))
        .replace("{provider_data_ingester}", f"{camel_provider}DataIngester")
        .replace("{endpoint}", endpoint)
        .replace("{provider_configuration}", provider_configuration)
    )

    return script_string
```

obulat replied:
> UTF-8 is the default encoding,

Actually, it's the default for the platform (https://docs.python.org/3/library/functions.html#open). I remember having problems reading files when I was learning Python on Windows.

AetherUnbound commented on openverse_catalog/templates/create_provider_ingester.py:
```diff
@@ -0,0 +1,180 @@
+    template_path: Path, provider: str, endpoint: str, media_types: list[str]
+):
+    with template_path.open("r", encoding="utf-8") as template:
+        camel_provider = inflection.camelize(provider)
+        screaming_snake_provider = inflection.underscore(provider).upper()
```
Neat!

obulat replied:
I've never heard about screaming snake case before 😆

AetherUnbound commented on openverse_catalog/templates/create_provider_ingester.py:
```diff
@@ -0,0 +1,180 @@
+    media_types: list[str],
+    name: str,
+):
+    with target.open("w", encoding="utf-8") as target_file:
+        filled_template = _get_filled_template(
+            template_path, provider, endpoint, media_types
+        )
+        target_file.write(filled_template)
+        print(f"{name + ':':<18} {target.relative_to(PROJECT_PATH)}")
```
Same note here for [`Path::write_text`](https://docs.python.org/3/library/pathlib.html#pathlib.Path.write_text):

```suggestion
    filled_template = _get_filled_template(
        template_path, provider, endpoint, media_types
    )
    target.write_text(filled_template)
    print(f"{name + ':':<18} {target.relative_to(PROJECT_PATH)}")
```

AetherUnbound commented on openverse_catalog/templates/create_provider_ingester.py:
```diff
@@ -0,0 +1,180 @@
+            template_path, provider, endpoint, media_types
+        )
+        target_file.write(filled_template)
+        print(f"{name + ':':<18} {target.relative_to(PROJECT_PATH)}")
```
Do you think we could use `logging` for messages throughout this module? A-la [the DAG documentation generation script](https://github.com/WordPress/openverse-catalog/blob/main/openverse_catalog/utilities/dag_doc_gen/dag_doc_generation.py#L23). Definitely don't have to, I just tend to like seeing the timestamps 😄

stacimc replied:
I tried briefly and had some trouble with logs being swallowed. I can debug when I have time, but this should also have very simple logs hopefully

AetherUnbound commented on openverse_catalog/templates/create_provider_ingester.py:
```diff
@@ -0,0 +1,180 @@
+def fill_template(provider, endpoint, media_types):
+    print(f"Creating files in {REPO_PATH}")
+
+    dags_path = TEMPLATES_PATH.parent / "dags" / "providers"
```
Can we use this here?
```suggestion
    dags_path = PROJECT_PATH / "dags" / "providers"
```

stacimc replied:
No, currently it would need to be REPO_PATH. But I think the use of PROJECT/REPO is wrong, I will update that.

AetherUnbound commented on openverse_catalog/templates/create_provider_ingester.py:
```diff
@@ -0,0 +1,180 @@
+
+
+def main():
+    parser = argparse.ArgumentParser(
```
Just a note, definitely doesn't need a change: Airflow comes with [Click](https://click.palletsprojects.com/en/8.1.x/), so we could use that here for these cases if this becomes more cumbersome 🙂

AetherUnbound commented on openverse_catalog/templates/template_provider.py_template:
```diff
@@ -0,0 +1,163 @@
+
+Content Provider:       {provider}
+
+ETL Process:            Use the API to identify all CC licensed images.
+
+Output:                 TSV file containing the images and the
+                        respective meta-data.
```
Should this reference the appropriate media? (rather than just "images")

AetherUnbound commented on openverse_catalog/templates/template_provider.py_template:
```diff
@@ -0,0 +1,163 @@
+                "limit": self.batch_limit,
+                "cc": 1,
+                "offset": 0,
+                "api_key": Variable.get("API_KEY_{screaming_snake_provider}")
```
😂 🐍 📣

AetherUnbound commented on openverse_catalog/templates/template_provider.py_template:
```diff
@@ -0,0 +1,163 @@
+    def get_media_type(self, record: dict):
+        # For a given record json, return the media type it represents.
+        # TODO: Update based on your API. TIP: May be hard-coded if the provider only
+        # returns records of one type.
```
Maybe include an example in this comment of what it could look like if we were just returning the constant?

AetherUnbound commented on tests/templates/test_create_provider_ingester.py:
```diff
@@ -0,0 +1,110 @@
+    "media_types_str, expected_types",
+    [
+        # Just image
+        (
+            [
+                "image",
+            ],
+            [
+                "image",
+            ],
+        ),
```
FWIW I believe if you remove the trailing comma from the entries, `black` should fit this all on one line  (same for the other entries below):
```suggestion
        (["image"], ["image"]),
```

AetherUnbound replied:
It's such an odd quirk of `black`, I wasn't aware of it for the longest time 😅

AetherUnbound commented:
Oh also, +1 to Zack's comment/request for a `just` recipe for running the provider creation script!

obulat commented on openverse_catalog/docs/data_models.md:
```diff
@@ -0,0 +1,57 @@
+| *foreign_identifier* | Unique identifier for the record on the source site. |
+| *thumbnail_url* | Direct link to a thumbnail-sized version of the record. |
+| *filesize* | Size of the main file in bytes. |
+| *filetype* | The filetype of the main file, eg. 'mp3', 'jpg', etc. |
```
Should we add something about the fact that filetype can be extracted from the file URL?
https://github.com/WordPress/openverse-catalog/blob/4a9c008a150b2fc5a88b8f9c64f5bd5beef66ac9/openverse_catalog/dags/common/storage/media.py#L293-L304

obulat commented on openverse_catalog/docs/provider_data_ingester_faq.md:
```diff
@@ -0,0 +1,142 @@
+
+**Example**: You're pulling data from a Museum database, and each "record" in a batch contains multiple photos of a single physical object.
+
+**Solution**: The `get_record_data` method takes a `data` object representing a single record from the provider API. Typically, it extracts required data and returns it as a single dict. However, it can also return a **list of dictionaries** for cases like the one described, where multiple Openverse records can be extracted.
```
It would be nice to add links to the examples of each of the solutions.

stacimc replied:
I wanted to do this but was worried about code drift and making this document difficult to maintain 🤔

AetherUnbound replied:
Ooo, that's a really good point!

obulat replied:
The problem with code drift and maintainability is really a good point :( I wonder if we could automate it somehow, add a link to some regex search string that would lead to the example of, say, overridden method in one of the solutions...  

Not necessary for this PR, but would be really nice to have. I usually learn by example and links to the examples of the solutions would be really helpful for the way _I_ learn :)

stacimc commented on openverse_catalog/templates/template_provider.py_template:
```diff
@@ -0,0 +1,163 @@
+def main():
+    # Allows running ingestion from the CLI without Airflow running for debugging
+    # purposes.
+    logger.info("Begin: {provider} data ingestion")
```
This is a really good point, because we're doing this in a lot of our `main` methods -- we should go back and remove it from them as well!

stacimc commented on openverse_catalog/docs/adding_a_new_provider.md:
```diff
@@ -0,0 +1,90 @@
+
+At a high level the steps are:
+
+1. `generate_filename`: Generates a TSV filename used in later steps
```
I love this!

stacimc commented on openverse_catalog/docs/adding_a_new_provider.md:
```diff
@@ -0,0 +1,90 @@
+
+### Add a `ProviderWorkflow` configuration class
+
+Now that you have an ingester class, you're ready to wire up a provider DAG in Airflow to automatically pull data and load it into our Catalog database. This is as simple as defining a `ProviderWorkflow` configuration dataclass and adding it to the `PROVIDER_WORKFLOWS` list in [`provider_workflows.py`](../dags/providers/provider_workflows.py). Our DAG factories will pick up the configuration and  ...
```
Ah, I thought I had removed that :) Totally agreed.

stacimc commented on openverse_catalog/docs/data_models.md:
```diff
@@ -0,0 +1,57 @@
+
+| field name | description |
+| --- | --- |
+| *foreign_identifier* | Unique identifier for the record on the source site. |
```
I really wondered about this! Obviously most of our provider scripts are already returning `None` early if the foreign id is not populated, but not all of them (Europeana doesn't), and none of the media stores do any validation. The table itself appears to `default uuid_generate_v4()`.

I ended up leaving it off because previous docs excluded it from the required fields as well 🤷‍♀️ I think it makes sense though to add it as required for _future_ providers, and we can decide if there are any more action items in the data normalization project.

stacimc commented:
Thank you so much for all the feedback! I've also added a `just` recipe, so we can now run:

```
just add-provider "Foo Museum" https://foo.museum.org/api/v1 image audio
```

It does not use the `run` command since I don't think there's any reason to ever run this using the webserver image, but it does save folks from needing to find the path to the script, and it keeps all our commands in one place :)

AetherUnbound commented on openverse_catalog/templates/template_provider.py_template:
```diff
@@ -0,0 +1,164 @@
+        #
+        # If a required field is missing, return early to prevent unnecesary
+        # processing.
+        if (foreign_identifier = data.get("foreign_id")) is None:
```
Small syntax issue!
```suggestion
        if (foreign_identifier := data.get("foreign_id")) is None:
```

stacimc replied:
🤦‍♀️ 😄

AetherUnbound reviewed (approved):
How exciting!! 💃🏼 I tried this out locally and it ran great. There's just one syntax issue in the generated script, otherwise this is good to go IMO

```
j add-provider "The Word For World Is Forest" http://the.forest image
python3 openverse_catalog/templates/create_provider_ingester.py "The Word For World Is Forest" "http://the.forest" -m image
Creating files in /home/madison/git-a8c/openverse-catalog
API script:        openverse_catalog/dags/providers/provider_api_scripts/the_word_for_world_is_forest.py
API script test:   tests/dags/providers/provider_api_scripts/test_the_word_for_world_is_forest.py

NOTE: You will also need to add a new ProviderWorkflow dataclass configuration to the PROVIDER_WORKFLOWS list in `openverse-catalog/dags/providers/provider_workflows.py`.
```
